### DISCUSSION TITLE
unity's world coordinate is left-hand, did you do something on the mesh and shader?

### DISCUSSION THREAD

**Discussion by WormCoder** [julienkay/MobileNeRF-Unity-Viewer#10] [Q&A] [answered]
the author‘s original code by WebGL, which is right-hand coordinate system

**Comment by julienkay** [comment, marked as answer]
Importing the .obj file is done using Unity's standard Asset Import Pipeline, which handles loading data from .obj in the correct way. So the mesh doesn't need any manual conversion, because Unity does it for you.

In the shader I do perform some coordinate system conversion. This gets a bit messy to explain, because the original code itself flips y-and z, but only for synthetic and unbounded 360° scenes (can be seen here https://github.com/google-research/jax3d/tree/main/jax3d/projects/mobilenerf if you diff the three view_*.html files). They do it inside the evaluateNetwork() function in the fragment shader.

In my Unity viewer, I opted to do that conversion in the vertex shader instead, where I also apply the conversion to a left handed coordinate system at the same time. This happens in the shader generation part where I'm replacing the AXIS_SWIZZLE keyword in the shader template:
https://github.com/julienkay/MobileNeRF-Unity-Viewer/blob/917704e5e82710420f8367aaa504167a17ac7a5b/Assets/Editor/MobileNeRFImporter.cs#L550-L552

Hope this helps.

**Reply by WormCoder** [reply]
Thanks for the explanation.

in your code for Synthetic 360° scenes:

// Synthetic 360° scenes
                return "o.rayDirection.xz = -o.rayDirection.xz;" +
                       "o.rayDirection.xyz = o.rayDirection.xzy;";

the sencond line  flips y-and z as the original webgl code.
but the first line makes x、z negative, could you explain the reason?

**Comment by julienkay** [comment]
This negation also happens as part of the coordinate conversion the original code does based on which type of scene it is. So it's actually not just switching y-z.

This can be seen if you compare the following code snippets:
https://github.com/google-research/jax3d/blob/7b81812661aed362e0587ac34fcfaf5cf624f9a0/jax3d/projects/mobilenerf/view_unbounded.html#L116
https://github.com/google-research/jax3d/blob/7b81812661aed362e0587ac34fcfaf5cf624f9a0/jax3d/projects/mobilenerf/view_forwardfacing.html#L116

One is:
`(j == 9) ? -viewdir.b : viewdir.g); //switch y-z axes`

the other is:
`(j == 9) ? viewdir.g : viewdir.b);`

As I've said I've opted to combine these conversions and do them in a single place, but I can see how that might have made things a bit more confusing.

**Comment by julienkay** [comment]
To be more clear, the negation of z is to match the original code.
The negation of x is to convert to Unity's left handed coordinate system.

**Reply by WormCoder** [reply]
thanks for your reply.
I thought Unity's world coordinate is Y-up left-hand, so we should make z negative to convert to Unity's left handed coordinate system.
maybe I'm wrong :)
